Typeinterval part2

Dr-Irv opened the pull request:
Is on top of https://github.com/pandas-dev/pandas/pull/46080

I believe the typing for `Interval` here is correct as it passes the tests created here:
https://github.com/microsoft/python-type-stubs/pull/167/

Leaving this as draft status as it requires #46080 to be approved.

Or a decision to go with #44922, in which case we kill #46080 and this one.

Dr-Irv committed:
fix column_arrays for array manager

Dr-Irv committed:
merge with upstream/main

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main'

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main'

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main'

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main'

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main'

Dr-Irv committed:
fix up typing to eventually support Interval typing

Dr-Irv committed:
fix imports in core/tools/datetimes.py

Dr-Irv committed:
fix time in odfreader

Dr-Irv committed:
pandas/core/arrays/masked.py

Dr-Irv committed:
use cast instead of new code in odfreader

Dr-Irv committed:
fix odfreader

Dr-Irv committed:
interval pyi

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,134 @@
+    @property
+    def open_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def mid(self: IntervalMixinProtocol) -> float: ...
```
`mid` should be generic. `length` should also be here, it is implemented in `IntervalMixin`, not in `Interval` (this will craete some [headaches](https://github.com/pandas-dev/pandas/pull/44922/files#diff-0a83f799817ae13154c34067c21d5a126efb6b433ec0860d88140a85751f8735R93) - because `IntervalTree` also inherits from  `IntervalMixin`).

I will not work on my PR for the next days, hopefully yours makes it. Later this week I have more time to pick up my PR if needed.

Dr-Irv replied:
fixed in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,134 @@
+    @property
+    def mid(self: IntervalMixinProtocol) -> float: ...
+    @property
+    def is_empty(self: IntervalMixinProtocol) -> bool: ...
```
Can you explain why you need to add `IntervalMixinProtocol` everywhere? I'm actually surprised this is actually allowed (any class implements this empty protocol, when using TypeVars pyright is picky about that, see https://github.com/microsoft/pyright/issues/3088#issuecomment-1046335018)

Dr-Irv replied:
YOu're right.  Moved it to be in the class definition of `IntervalMixin`

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,134 @@
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
+    @overload
+    def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> Timedelta: ...
```
I assume Interval allows sub-classes of int/float/Timestamp/Timedelta: methods might(?) then also return the sub-classed types (not sure whether that is worth the trouble).

Dr-Irv replied:
I don't know if the cython implementation will allow that, and I say let's not worry about it.

twoertwein replied:
Do mypy&pyright accept this lonely `OrderableTimesT`, typically they need to be used somewhere else as well.

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,134 @@
+        self,
+        left: OrderableT,
+        right: OrderableT,
+        closed: Union[str, Literal["left", "right", "both", "neither"]] = ...,
```
remove `str`, wouldn't need the literals otherwise.

Dr-Irv replied:
OK

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,134 @@
+class IntervalTree(IntervalMixin):
+    def __init__(
+        self,
+        left: np.ndarray,
```
You can make this class generic to keep track of the dtype

twoertwein replied:
Probably should be generic, because `IntervalMixin` needs to be generic (for length/mid).

Dr-Irv replied:
Not sure that is necessary - we're OK without it

twoertwein replied:
I see you found a way to avoid `IntervalMixin` being generic :)

Since the implementation [appears](https://github.com/pandas-dev/pandas/blob/ea0184e339b7dbf27eec53e6c01c6514bda1fc83/pandas/_libs/intervaltree.pxi.in#L61) to allow ndarrays of different types, it should be fine not to track the dtype of the ndarray.

Dr-Irv replied:
later, you indicated to remove the generic, so resolving this

Dr-Irv committed:
use cast in odfreader

Dr-Irv committed:
use cast in odfreader

Dr-Irv committed:
fixes for mid, length to mixin, change use of protocol

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    @overload
+    def __get__(self, instance: Interval[Timedelta], owner: Any) -> Timedelta: ...
+    @overload
+    def __get__(self, instance: Interval[Timestamp], owner: Any) -> Timestamp: ...
```
You can probably collapse the last two overloads into one (which also takes care of sub-classes):

```py
@overload
def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> OrderableTimesT: ...
```

You will also need an overload for IntervalTree (for `mid/length`).

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
```
Can remove the `else`, pandas requires 3.8 :)

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    @overload
+    def __get__(self, instance: Interval[Timestamp], owner: Any) -> Timestamp: ...
+
+class IntervalMixin(IntervalMixinProtocol):
```
`IntervalMixinProtocol` can probably be removed

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+
+OrderableScalarT = TypeVar("OrderableScalarT", int, float)
+OrderableTimesT = TypeVar("OrderableTimesT", Timestamp, Timedelta)
+OrderableT = TypeVar("OrderableT", int, float, Timestamp, Timedelta)
```
Should probably prefix the TypeVars with `_` to not mark them as public.

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    length: _LengthDescriptor
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...): ...
```
returns `None`

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    @property
+    def right(self: Interval[OrderableT]) -> OrderableT: ...
+    @property
+    def closed(self) -> str: ...
```
Could be `Literal["left", "right", "both", "neither"]`

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    @overload
+    def __contains__(self: Interval[int], key: Union[int, float]) -> bool: ...
+    @overload
+    def __contains__(self: Interval[float], key: Union[int, float]) -> bool: ...
```
Could combine the last two overloads in
```py
@overload
def __contains__(self: Interval[float] | Interval[float], key: int | float) -> bool: ...
```

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+        self: Interval[OrderableTimesT], y: Timedelta
+    ) -> Interval[OrderableTimesT]: ...
+    @overload
+    def __add__(self: Interval[int], y: int) -> Interval[int]: ...
```
Not sure: might need to add the reverse operators (radd, rsub, ...) as well - would only add them if they fix typing errors in the pandas code base.

Dr-Irv replied:
Probably should add for external typing reasons.  Will do it in next commit.

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+else:
+    from typing_extensions import Literal
+
+VALID_CLOSED: frozenset
```
`frozenset[str]`

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    ) -> Interval[float]: ...
+    def overlaps(self: Interval[OrderableT], other: Interval[OrderableT]) -> bool: ...
+
+def intervals_to_interval_bounds(intervals: np.ndarray, validate_closed: int = ...): ...
```
I think this returns `tuple[np.ndarray, np.ndarray, str]`

`validate_closed` should probably be `bool` (based on the doc-string).

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+        left: np.ndarray,
+        right: np.ndarray,
+        closed: Literal["left", "right", "both", "neither"] = ...,
+    ): ...
```
`leaf_size: int = ...` is missing

Dr-Irv replied:
fixed in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+        closed: Literal["left", "right", "both", "neither"] = ...,
+    ): ...
+    def get_indexer(self, target) -> np.ndarray: ...
+    def get_indexer_non_unique(self, target) -> np.ndarray: ...
```
returns `tuple[npt.NDArray[np.intp], npt.NDArray[np.intp]]`

Dr-Irv replied:
done in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+        right: np.ndarray,
+        closed: Literal["left", "right", "both", "neither"] = ...,
+    ): ...
+    def get_indexer(self, target) -> np.ndarray: ...
```
return value can be slightly narrower: `npt.NDArray[np.intp]`

Dr-Irv replied:
fixed in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,141 @@
+    def get_indexer_non_unique(self, target) -> np.ndarray: ...
+    _na_count: int
+    @property
+    def is_overlapping(self) -> bool: ...
```
missing methods:
```py
    @property
    def is_monotonic_increasing(self) -> bool: ...
    def clear_mapping(self) -> None:
```

Dr-Irv replied:
fixed in next commit

twoertwein commented:
Looks pretty good @Dr-Irv! I'll look at part 1 tomorrow.

Dr-Irv commented:
> Looks pretty good @Dr-Irv! I'll look at part 1 tomorrow.

Thanks!

Dr-Irv committed:
misc cleanup from twoertwein

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,179 @@
+    intervals: np.ndarray, validate_closed: bool = ...
+) -> Tuple[np.ndarray, np.ndarray, str]: ...
+
+class IntervalTree(IntervalMixin, Generic[_OrderableT]):
```
I don't think you need this generic here.

In _MidDescriptor you can probably simply do

```py
    @overload
    def __get__(
        self, instance: IntervalTree, owner: Any
    ) -> np.ndarray: ...
```
and the same for `length`

Dr-Irv replied:
done in next commit

Dr-Irv committed:
remove generic from IntervalTree

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main' into typeinterval_part1

Dr-Irv committed:
clean up based on comments.  simplify python_parser changes. Introduce IntervalClosedType

Dr-Irv committed:
Merge branch 'typeinterval_part1' into typeinterval_part2

jreback commented:
other PR merged so pls rebase

Dr-Irv committed:
Merge remote-tracking branch 'upstream/main' into typeinterval_part2

Dr-Irv commented:
> other PR merged so pls rebase

all set and all green

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,175 @@
+    @overload
+    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
```
I think these two overloads could be:

`def __get__(self, instance: Interval[_OrderableScalarT], owner: Any) -> _OrderableScalarT:`

Dr-Irv replied:
fixed in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,175 @@
+    @property
+    def right(self: Interval[_OrderableT]) -> _OrderableT: ...
+    @property
+    def closed(self) -> Literal["left", "right", "both", "neither"]: ...
```
Could use `_typing.IntervalClosedType` here and in a few more places

Dr-Irv replied:
fixed in next commit

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,175 @@
+
+def intervals_to_interval_bounds(
+    intervals: np.ndarray, validate_closed: bool = ...
+) -> Tuple[np.ndarray, np.ndarray, str]: ...
```
Can use `tuple` Is pyupgrade run on pyi files?

Dr-Irv replied:
guess not - for someone else to fix!

twoertwein commented on pandas/_libs/interval.pyi:
```diff
@@ -0,0 +1,175 @@
+    def get_indexer(self, target) -> npt.NDArray[np.intp]: ...
+    def get_indexer_non_unique(
+        self, target
+    ) -> Tuple[npt.NDArray[np.intp], npt.NDArray[np.intp]]: ...
```
`tuple`

Dr-Irv replied:
fixed in next commit

Dr-Irv committed:
Feedback from twoertwein
- Use IntervalClosedType
- Consolidate __get__ in LengthDescriptor
- Use tuple instead of Tuple

twoertwein reviewed (approved):
Great @Dr-Irv  looks go to me!

jreback commented:
thanks @Dr-Irv keep em coming!
